ngraph/core/reference/include/ngraph/runtime/reference/cum_sum.hpp

```diff
@@ -19,115 +22,134 @@ namespace ngraph
     {
         namespace reference
         {
-            template <typename T, typename P>
-            void cumsum(const T* arg,
-                        const P* axis_tensor,
-                        T* out,
-                        const Shape& tensor_shape,
-                        const bool exclusive,
-                        const bool reverse)
+            namespace details
             {
-                NGRAPH_SUPPRESS_DEPRECATED_START
-                CoordinateTransform temp_transform(tensor_shape);
-                for (const Coordinate& output_coord : temp_transform)
+                template <typename InputIterator, typename OutputIterator>
+                void flat_cumsum(InputIterator it_input_begin,
+                                 InputIterator it_input_end,
+                                 OutputIterator it_output_begin,
+                                 const bool exclusive)
                 {
-                    out[temp_transform.index(output_coord)] = 0;
+                    if (exclusive)
+                    {
+                        --it_input_end;
+                        ++it_output_begin;
+                    }
+                    std::partial_sum(it_input_begin, it_input_end, it_output_begin);
                 }
+                template <typename T>
+                void loop_cumsum(std::vector<T>& output_data,
+                                 const bool exclusive,
+                                 const bool reverse,
+                                 const Shape& shape,
+                                 const T* data_ptr,
+                                 const size_t axis_dim_size)
+                {
+                    const auto slices_count =
+                        shape_size(Shape(shape.begin(), shape.begin() + shape.size() - 1));
+                    auto axis_dim_counter = 0;
+                    for (auto i = 0; i < slices_count; ++i)
+                    {
+                        std::vector<T> input_data(data_ptr, data_ptr + axis_dim_size);
+                        std::vector<T> output_tmp_data(axis_dim_size, 0);
+                        if (reverse)
+                        {
+                            details::flat_cumsum(input_data.rbegin(),
+                                                 input_data.rend(),
+                                                 output_tmp_data.rbegin(),
+                                                 exclusive);
+                        }
+                        else
+                        {
+                            details::flat_cumsum(input_data.begin(),
+                                                 input_data.end(),
+                                                 output_tmp_data.begin(),
+                                                 exclusive);
+                        }
+                        std::move(output_tmp_data.begin(),
```

Comment:
There no advantage when `std::move` is used instead of `std::copy` for fundamental types.

Reply:
I see, but does it mean that usage of `std::move` here is wrong?

Reply:
Probably it doesn't bring any negative impact.
This is my perspective: we are working on memory buffers and I'll preferer here `std::memcpy` or `std::copy`.